src/lambda_handler.py

```diff
@@ -1,89 +1,207 @@
-    numerator: str = event.get("numerator", None)
-    denominator: str = event.get("denominator", None)
 
-    if denominator is None or numerator is None:
-        error_msg = "Request must include both a numerator and a denominator."
-        logging.error(error_msg)
-        failed_task(result, error_msg)
-    else:
-        try:
-            variable_error_msg = "numerator: %s, denominator: %s"
-            result["result"] = example_div(int(numerator), int(denominator))
-        except ValueError:
-            error_msg = "The provided values must be integers."
-            logging.error(error_msg)
-            logging.error(variable_error_msg, numerator, denominator)
-            failed_task(result, error_msg)
-        except ZeroDivisionError:
-            error_msg = "The denominator cannot be zero."
-            logging.error(error_msg)
-            logging.error(variable_error_msg, numerator, denominator)
-            failed_task(result, error_msg)
-
-    return result
-
-
-def handler(event, context) -> dict[str, Optional[str]]:
+async def process_cve_json(json_stream) -> None:
+    """Process the provided CVEs JSONs and update the database with its contents."""
+    data = json.load(json_stream)
+
+    if data.get("CVE_data_type") != "CVE":
+        raise ValueError("JSON does not look like valid NVD CVE data.")
+
+    tasks = [asyncio.create_task(process_nvd(cve)) for cve in data.get("CVE_Items", [])]
+
+    for task in asyncio.as_completed(tasks):
+        await task
+
+
+async def process_urls(cve_urls, db) -> None:
+    """Initialize beanie ODM and begin processing CVE URLs."""
+    await init_beanie(database=motor_client[db], document_models=[CVEDoc])
+    # Async for loop that processes each of the CVE URLs individually
+
+    for cve_url in cve_urls:
+        # We disable the bandit blacklist for the urllib.request.urlopen() function
+        # because the URL is either the defaul (safe) URL or one provided in the
+        # Lambda configuration so we can assume it is safe.
+        retries = 0
+        attempts = 10
+
+        print("Connecting to socket from: ", cve_url)
+
+        while True:
+            try:
+                socket = urllib.request.urlopen(cve_url)  # nosec B310
+                break
+            except (urllib.error.HTTPError, urllib.error.URLError) as err:
+                logging.debug(
+                    "Encountered a(n) %s exception while attempting to open socket from '%s'",
+                    type(err).__name__,
+                    cve_url,
+                )
+                retries += 1
+                await asyncio.sleep(5)
+                if retries <= attempts:
+                    logging.warning(
+                        "Performing retry %d/%d for '%s'", retries, attempts, cve_url
+                    )
+                else:
+                    raise err
+
+        buf = BytesIO(socket.read())
+        f = gzip.GzipFile(fileobj=buf)
+        await process_cve_json(f)
```

Comment:
You went through the trouble to use `async` throughout this PR, but it seems like your use of `await` is almost always such that it forces synchronous behavior.  Did you want to change the logic to schedule _all_ the `process_cve_json(f)` tasks and then wait for them _all_ to finish?  That would allow more than one to run concurrently.

Reply:
What is the reason why we wouldn't want to pull in multiple years of CVEs concurrently?

Reply:
@dav3r and @jsf9k I had to make a few fairly big changes to how we were reading from the URLs to get the CVEs to run concurrently. Before we were `url.request.urlopen(cve_url)` but when trying to group the reading tasks together I kept running into an error `IncompleteRead(xxxxxx bytes read, xxxxxx more expected)`. I switched to using `aiohttp` to read from the URLs you will see it here [Lines 141 - 150](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R141-R150) and also here [Lines 157 - 188](https://github.com/cisagov/cyhy-nvdsync-lambda/pull/2/commits/e6397c680698de8835b0629e200b28fe2829df5b#diff-4655711b74e9f79a2566fdbac7451f4dcc265a12d3ad743e519353283898a155R157-R188).